File: src/main/java/org/mtransit/android/ui/fragment/MapFragment.java
package org.mtransit.android.ui.fragment;

import android.app.Activity;
import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Color;
import android.location.Location;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.Loader;

import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;

import org.mtransit.android.R;
import org.mtransit.android.common.IContext;
import org.mtransit.android.commons.BundleUtils;
import org.mtransit.android.commons.LocationUtils;
import org.mtransit.android.commons.MTLog;
import org.mtransit.android.commons.PreferenceUtils;
import org.mtransit.android.commons.TaskUtils;
import org.mtransit.android.data.DataSourceProvider;
import org.mtransit.android.data.DataSourceType;
import org.mtransit.android.di.Injection;
import org.mtransit.android.provider.permission.LocationPermissionProvider;
import org.mtransit.android.task.MTCancellableFragmentAsyncTask;
import org.mtransit.android.task.MapPOILoader;
import org.mtransit.android.ui.ActionBarController;
import org.mtransit.android.ui.MTActivityWithLocation;
import org.mtransit.android.ui.MTDialog;
import org.mtransit.android.ui.view.MapViewController;
import org.mtransit.android.util.CrashUtils;
import org.mtransit.android.util.LoaderUtils;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

public class MapFragment extends ABFragment implements
		LoaderManager.LoaderCallbacks<Collection<MapViewController.POIMarker>>,
		MTActivityWithLocation.UserLocationListener,
		IContext,
		MapViewController.MapListener {

	private static final String LOG_TAG = MapFragment.class.getSimpleName();

	@NonNull
	@Override
	public String getLogTag() {
		return LOG_TAG;
	}

	private static final String TRACKING_SCREEN_NAME = "Map";

	@NonNull
	@Override
	public String getScreenName() {
		return TRACKING_SCREEN_NAME;
	}

	private static final String EXTRA_INITIAL_LOCATION = "extra_initial_location";
	private static final String EXTRA_SELECTED_UUID = "extra_selected_uuid";
	private static final String EXTRA_INCLUDE_TYPE_ID = "extra_include_type_id";

	@NonNull
	public static MapFragment newInstance(@Nullable Location optInitialLocation, @Nullable String optSelectedUUID, @Nullable Integer optIncludeTypeId) {
		MapFragment f = new MapFragment();
		Bundle args = new Bundle();
		if (optInitialLocation != null) {
			args.putParcelable(EXTRA_INITIAL_LOCATION, optInitialLocation);
		}
		if (!TextUtils.isEmpty(optSelectedUUID)) {
			args.putString(EXTRA_SELECTED_UUID, optSelectedUUID);
		}
		if (optIncludeTypeId != null) {
			args.putInt(EXTRA_INCLUDE_TYPE_ID, optIncludeTypeId);
			f.includedTypeId = optIncludeTypeId;
		}
		f.setArguments(args);
		return f;
	}

	@NonNull
	private final MapViewController mapViewController =
			new MapViewController(LOG_TAG, null, this, true, true, true, false, false, false, 64, false, true, true, false, true);

	@NonNull
	private final LocationPermissionProvider locationPermissionProvider;

	public MapFragment() {
		super();
		this.locationPermissionProvider = Injection.providesLocationPermissionProvider();
	}

	@Override
	public void onAttach(@NonNull Activity activity) {
		super.onAttach(activity);
		this.mapViewController.setLocationPermissionGranted(this.locationPermissionProvider.permissionsGranted(this));
		this.mapViewController.onAttach(activity);
	}

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setHasOptionsMenu(true);
		restoreInstanceState(savedInstanceState, getArguments());
		this.mapViewController.onCreate(savedInstanceState);
	}

	@Nullable
	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		super.onCreateView(inflater, container, savedInstanceState);
		View view = inflater.inflate(R.layout.fragment_map, container, false);
		this.mapViewController.onCreateView(view, savedInstanceState);
		return view;
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		setupView(view);
		this.mapViewController.onViewCreated(view, savedInstanceState);
	}

	private void setupView(@SuppressWarnings("unused") View view) {
		// DO NOTHING
	}

	private void restoreInstanceState(Bundle... bundles) {
		Location newInitialLocation = BundleUtils.getParcelable(EXTRA_INITIAL_LOCATION, bundles);
		if (newInitialLocation != null) {
			this.mapViewController.setInitialLocation(newInitialLocation);
		}
		String newSelectedUUID = BundleUtils.getString(EXTRA_SELECTED_UUID, bundles);
		if (!TextUtils.isEmpty(newSelectedUUID)) {
			this.mapViewController.setInitialSelectedUUID(newSelectedUUID);
		}
		Integer newIncludedTypeId = BundleUtils.getInt(EXTRA_INCLUDE_TYPE_ID, bundles);
		if (newIncludedTypeId != null) {
			this.includedTypeId = newIncludedTypeId;
		}
		this.mapViewController.setTag(getLogTag());
	}

	@Override
	public void onResume() {
		super.onResume();
		View view = getView();
		if (this.modulesUpdated) {
			if (view != null) {
				view.post(() -> {
					if (MapFragment.this.modulesUpdated) {
						onModulesUpdated();
					}
				});
			}
		}
		this.mapViewController.onResume();
		hasFilterTypeIds(); // triggers markers loading if necessary
		this.mapViewController.showMap(view);
		if (getActivity() != null) {
			onUserLocationChanged(((MTActivityWithLocation) getActivity()).getLastLocation());
		}
	}

	private boolean modulesUpdated = false;

	@Override
	public void onModulesUpdated() {
		this.modulesUpdated = true;
		if (!isResumed()) {
			return;
		}
		if (hasFilterTypeIds()) {
			resetTypeFilterIds();
			initFilterTypeIdsAsync();
		}
		this.modulesUpdated = false; // processed
	}

	private Location userLocation;

	@Override
	public void onUserLocationChanged(@Nullable Location newLocation) {
		if (newLocation == null) {
			return;
		}
		if (this.userLocation == null) {
			this.mapViewController.clearMarkers(); // previous marker for unspecific location
			this.loadedLatLngBounds = null; // reset loaded marker area
			this.mapViewController.setLocationPermissionGranted(this.locationPermissionProvider.permissionsGranted(this));
		}
		if (this.userLocation == null || LocationUtils.isMoreRelevant(getLogTag(), this.userLocation, newLocation)) {
			this.userLocation = newLocation;
		}
		this.mapViewController.onUserLocationChanged(newLocation);
	}

	private static final int POIS_LOADER = 0;

	@NonNull
	@Override
	public Loader<Collection<MapViewController.POIMarker>> onCreateLoader(int id, @Nullable Bundle args) {
		switch (id) {
		case POIS_LOADER:
			return new MapPOILoader(getContext(), getFilterTypeIdsOrNull(), this.loadingLatLngBounds, this.loadedLatLngBounds);
		default:
			CrashUtils.w(this, "Loader id '%s' unknown!", id);
			//noinspection ConstantConditions // TODO fix latter
			return null;
		}
	}

	@Override
	public void onLoaderReset(@NonNull Loader<Collection<MapViewController.POIMarker>> loader) {
	}

	@Override
	public void onLoadFinished(@NonNull Loader<Collection<MapViewController.POIMarker>> loader, @Nullable Collection<MapViewController.POIMarker> data) {
		if (this.loadingLatLngBounds == null) {
			return;
		}
		if (this.loadedLatLngBounds != null) {
			this.mapViewController.addMarkers(data);
			this.loadedLatLngBounds = this.loadingLatLngBounds;
			this.loadingLatLngBounds = null;
			this.mapViewController.hideLoading();
		} else {
			this.mapViewController.addMarkers(data);
			this.loadedLatLngBounds = this.loadingLatLngBounds;
			this.loadingLatLngBounds = null;
			this.mapViewController.showMap(getView());
		}
	}

	@Nullable
	private LatLngBounds loadingLatLngBounds = null;
	@Nullable
	private LatLngBounds loadedLatLngBounds = null;

	@Override
	public void onConfigurationChanged(@NonNull Configuration newConfig) {
		super.onConfigurationChanged(newConfig);
		this.mapViewController.onConfigurationChanged(newConfig);
	}

	@Override
	public void onMapClick(LatLng position) {
		// DO NOTHING
	}

	@Override
	public void onCameraChange(LatLngBounds latLngBounds) {
		if (latLngBounds == null) {
			return;
		}
		boolean loaded = this.loadedLatLngBounds != null //
				&& this.loadedLatLngBounds.contains(latLngBounds.northeast) //
				&& this.loadedLatLngBounds.contains(latLngBounds.southwest);
		boolean loading = this.loadingLatLngBounds != null //
				&& this.loadingLatLngBounds.contains(latLngBounds.northeast) //
				&& this.loadingLatLngBounds.contains(latLngBounds.southwest);
		if (!loaded && !loading) {
			this.mapViewController.showLoading();
			LoaderUtils.destroyLoader(this, POIS_LOADER); // cancel now
			if (this.loadingLatLngBounds != null) {
				if (!this.loadingLatLngBounds.contains(latLngBounds.northeast)) {
					this.loadingLatLngBounds = this.loadingLatLngBounds.including(latLngBounds.northeast);
				}
				if (!this.loadingLatLngBounds.contains(latLngBounds.southwest)) {
					this.loadingLatLngBounds = this.loadingLatLngBounds.including(latLngBounds.southwest);
				}
			} else if (this.loadedLatLngBounds != null) {
				this.loadingLatLngBounds = this.loadedLatLngBounds;
				if (!this.loadingLatLngBounds.contains(latLngBounds.northeast)) {
					this.loadingLatLngBounds = this.loadingLatLngBounds.including(latLngBounds.northeast);
				}
				if (!this.loadingLatLngBounds.contains(latLngBounds.southwest)) {
					this.loadingLatLngBounds = this.loadingLatLngBounds.including(latLngBounds.southwest);
				}
			} else {
				this.loadingLatLngBounds = latLngBounds;
				float factor = 1.0f;
				LatLngBounds bigLatLngBounds = this.mapViewController.getBigCameraPosition(getActivity(), factor);
				if (bigLatLngBounds != null) {
					this.loadingLatLngBounds = bigLatLngBounds;
				}
			}
			if (hasFilterTypeIds()) {
				LoaderUtils.restartLoader(this, POIS_LOADER, null, this);
			}
		}
	}

	@Override
	public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
		super.onCreateOptionsMenu(menu, inflater);
		inflater.inflate(R.menu.menu_map, menu);
	}

	private Integer includedTypeId = null;
	private Set<Integer> filterTypeIds = null;

	private void resetTypeFilterIds() {
		this.filterTypeIds = null; // reset
	}

	private Set<Integer> getFilterTypeIdsOrNull() {
		if (!hasFilterTypeIds()) {
			return null;
		}
		return this.filterTypeIds;
	}

	private boolean hasFilterTypeIds() {
		if (this.filterTypeIds == null) {
			initFilterTypeIdsAsync();
			return false;
		}
		return true;
	}

	private void initFilterTypeIdsAsync() {
		if (this.loadFilterTypeIdsTask != null && this.loadFilterTypeIdsTask.getStatus() == LoadFilterTypeIdsTask.Status.RUNNING) {
			return;
		}
		this.loadFilterTypeIdsTask = new LoadFilterTypeIdsTask(this);
		TaskUtils.execute(this.loadFilterTypeIdsTask);
	}

	@Nullable
	private LoadFilterTypeIdsTask loadFilterTypeIdsTask = null;

	private static class LoadFilterTypeIdsTask extends MTCancellableFragmentAsyncTask<Void, Void, Boolean, MapFragment> {

		@NonNull
		@Override
		public String getLogTag() {
			return MapFragment.class.getSimpleName() + ">" + LoadFilterTypeIdsTask.class.getSimpleName();
		}

		LoadFilterTypeIdsTask(@NonNull MapFragment mapFragment) {
			super(mapFragment);
		}

		@Nullable
		@Override
		protected Boolean doInBackgroundNotCancelledWithFragmentMT(@NonNull MapFragment mapFragment, @Nullable Void... params) {
			return mapFragment.initFilterTypeIdsSync();
		}

		@Override
		protected void onPostExecuteNotCancelledFragmentReadyMT(@NonNull MapFragment mapFragment, @Nullable Boolean result) {
			if (Boolean.TRUE.equals(result)) {
				mapFragment.applyNewFilterTypeIds();
			}
		}
	}

	private boolean initFilterTypeIdsSync() {
		if (this.filterTypeIds != null) {
			return false;
		}
		final Context context = getContext();
		if (context == null) {
			return false;
		}
		ArrayList<DataSourceType> availableTypes = filterTypes(DataSourceProvider.get(context).getAvailableAgencyTypes());
		Set<String> filterTypeIdStrings = PreferenceUtils.getPrefLcl( //
				context, PreferenceUtils.PREFS_LCL_MAP_FILTER_TYPE_IDS, PreferenceUtils.PREFS_LCL_MAP_FILTER_TYPE_IDS_DEFAULT);
		this.filterTypeIds = new HashSet<>();
		boolean hasChanged = false;
		if (filterTypeIdStrings != null) {
			for (String typeIdString : filterTypeIdStrings) {
				try {
					DataSourceType type = DataSourceType.parseId(Integer.parseInt(typeIdString));
					if (type == null) {
						hasChanged = true;
						continue;
					}
					if (!availableTypes.contains(type)) {
						hasChanged = true;
						continue;
					}
					this.filterTypeIds.add(type.getId());
				} catch (Exception e) {
					MTLog.w(this, e, "Error while parsing filter type ID '%s'!", typeIdString);
					hasChanged = true;
				}
			}
		}
		if (this.includedTypeId != null) {
			if (this.filterTypeIds.size() > 0 && !this.filterTypeIds.contains(this.includedTypeId)) {
				try {
					DataSourceType type = DataSourceType.parseId(this.includedTypeId);
					if (type == null) {
						// DO NOTHING
					} else if (!availableTypes.contains(type)) {
						// DO NOTHING
					} else {
						this.filterTypeIds.add(type.getId());
						hasChanged = true;
					}
				} catch (Exception e) {
					MTLog.w(this, e, "Error while parsing filter type ID '%s'!", this.includedTypeId);
					hasChanged = true;
				}
			}
			this.includedTypeId = null; // only once
		}
		if (hasChanged) { // old setting not valid anymore
			saveMapFilterTypeIdsSetting(false); // asynchronous
		}
		return this.filterTypeIds != null;
	}

	@NonNull
	private ArrayList<DataSourceType> filterTypes(@NonNull ArrayList<DataSourceType> availableTypes) {
		Iterator<DataSourceType> it = availableTypes.iterator();
		while (it.hasNext()) {
			if (!it.next().isMapScreen()) {
				it.remove();
			}
		}
		return availableTypes;
	}

	private void saveMapFilterTypeIdsSetting(boolean sync) {
		Set<Integer> filterTypeIds = getFilterTypeIdsOrNull();
		if (filterTypeIds == null) {
			return;
		}
		Set<String> newFilterTypeIdStrings = new HashSet<>();
		for (Integer filterTypeId : filterTypeIds) {
			newFilterTypeIdStrings.add(String.valueOf(filterTypeId));
		}
		PreferenceUtils.savePrefLcl(getContext(), PreferenceUtils.PREFS_LCL_MAP_FILTER_TYPE_IDS, newFilterTypeIdStrings, sync);
	}

	private void applyNewFilterTypeIds() {
		if (this.filterTypeIds == null) {
			return;
		}
		LoaderUtils.destroyLoader(this, POIS_LOADER); // cancel now
		final ActionBarController abController = getAbController();
		if (abController != null) {
			abController.setABTitle(this, getABTitle(requireContext()), true);
		}
		if (this.loadingLatLngBounds == null) {
			this.loadingLatLngBounds = this.loadedLatLngBounds; // use the loaded area
		}
		this.loadedLatLngBounds = null; // loaded with wrong filter
		this.mapViewController.clearMarkers();
		this.mapViewController.showMap(getView());
		if (this.loadingLatLngBounds != null) {
			LoaderUtils.restartLoader(this, POIS_LOADER, null, this);
		}
	}

	@Override
	public boolean onOptionsItemSelected(@NonNull MenuItem item) {
		switch (item.getItemId()) {
		case R.id.menu_filter:
			Set<Integer> filterTypeIds = getFilterTypeIdsOrNull();
			if (filterTypeIds == null) {
				return false;
			}
			ArrayList<CharSequence> typeNames = new ArrayList<>();
			ArrayList<Boolean> checked = new ArrayList<>();
			final ArrayList<Integer> typeIds = new ArrayList<>();
			final HashSet<Integer> selectedItems = new HashSet<>();
			ArrayList<DataSourceType> availableAgencyTypes = filterTypes(DataSourceProvider.get(getContext()).getAvailableAgencyTypes());
			for (DataSourceType type : availableAgencyTypes) {
				typeIds.add(type.getId());
				typeNames.add(getString(type.getPoiShortNameResId()));
				checked.add(filterTypeIds.size() == 0 || filterTypeIds.contains(type.getId()));
			}
			boolean[] checkedItems = new boolean[checked.size()];
			for (int c = 0; c < checked.size(); c++) {
				checkedItems[c] = checked.get(c);
				if (checkedItems[c]) {
					selectedItems.add(c);
				}
			}
			new MTDialog.Builder(requireActivity()) //
					.setTitle(R.string.menu_action_filter) //
					.setMultiChoiceItems( //
							typeNames.toArray(new CharSequence[0]), //
							checkedItems, //
							(dialog, which, isChecked) -> {
								if (isChecked) {
									selectedItems.add(which);
								} else {
									selectedItems.remove(which);
								}
							} //
					) //
					.setPositiveButton(android.R.string.ok, (dialog, which) -> {
						applyNewFilter(typeIds, selectedItems);
						dialog.dismiss();
					}) //
					.setNegativeButton(android.R.string.cancel, (dialog, which) ->
							dialog.dismiss()
					) //
					.setCancelable(true) //
					.create()
					.show();

			return true; // handled
		}
		return super.onOptionsItemSelected(item);
	}

	@Override
	public void onPause() {
		super.onPause();
		this.mapViewController.onPause();
	}

	@Override
	public void onSaveInstanceState(@NonNull Bundle outState) {
		this.mapViewController.onSaveInstanceState(outState);
		super.onSaveInstanceState(outState);
	}

	@Override
	public void onLowMemory() {
		super.onLowMemory();
		this.mapViewController.onLowMemory();
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();
		this.mapViewController.onDestroyView();
		this.loadedLatLngBounds = null;
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		this.mapViewController.onDestroy();
		TaskUtils.cancelQuietly(this.loadFilterTypeIdsTask, true);
	}

	@Override
	public void onDetach() {
		super.onDetach();
		this.mapViewController.onDetach();
	}

	@Override
	public CharSequence getABTitle(Context context) {
		StringBuilder sb = new StringBuilder(context.getString(R.string.map));
		sb.append(" (");
		Set<Integer> filterTypeIds = getFilterTypeIdsOrNull();
		if (filterTypeIds != null && filterTypeIds.size() > 0) {
			boolean hasType = false;
			for (Integer typeId : filterTypeIds) {
				if (typeId != null) {
					DataSourceType type = DataSourceType.parseId(typeId);
					if (type != null) {
						if (hasType) {
							sb.append(", ");
						}
						sb.append(getString(type.getAllStringResId()));
						hasType = true;
					}
				}
			}
		} else {
			sb.append(context.getString(R.string.all));
		}
		sb.append(")");
		return sb.toString();
	}

	@Override
	public Integer getABBgColor(Context context) {
		return Color.TRANSPARENT;
	}

	private void applyNewFilter(ArrayList<Integer> typeIds, HashSet<Integer> selectedItems) {
		boolean filterChanged = false;
		if (this.filterTypeIds == null) {
			return;
		}
		if (selectedItems.size() == 0 || selectedItems.size() == typeIds.size()) {
			if (this.filterTypeIds.size() > 0) {
				this.filterTypeIds.clear();
				filterChanged = true;
			}
		} else {
			for (int i = 0; i < typeIds.size(); i++) {
				Integer typeId = typeIds.get(i);
				if (selectedItems.contains(i)) {
					if (!this.filterTypeIds.contains(typeId)) {
						this.filterTypeIds.add(typeId);
						filterChanged = true;
					}
				} else {
					if (this.filterTypeIds.contains(typeId)) {
						this.filterTypeIds.remove(typeId);
						filterChanged = true;
					}
				}
			}
		}
		if (filterChanged) {
			saveMapFilterTypeIdsSetting(false); // asynchronous
			applyNewFilterTypeIds();
		}
	}
}

File: src/main/java/org/mtransit/android/ui/map/MapFragment.kt
@file:JvmName("MapFragment") // ANALYTICS
package org.mtransit.android.ui.map

import android.content.Context
import android.content.res.Configuration
import android.graphics.Color
import android.location.Location
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.core.os.bundleOf
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import dagger.hilt.android.AndroidEntryPoint
import org.mtransit.android.R
import org.mtransit.android.commons.LocationUtils
import org.mtransit.android.data.DataSourceType
import org.mtransit.android.data.POIManager
import org.mtransit.android.databinding.FragmentMapBinding
import org.mtransit.android.datasource.DataSourcesRepository
import org.mtransit.android.provider.permission.LocationPermissionProvider
import org.mtransit.android.ui.MTActivityWithLocation
import org.mtransit.android.ui.MTActivityWithLocation.UserLocationListener
import org.mtransit.android.ui.MTDialog
import org.mtransit.android.ui.fragment.ABFragment
import org.mtransit.android.ui.view.MapViewController
import org.mtransit.android.ui.view.common.isAttached
import javax.inject.Inject

@AndroidEntryPoint
class MapFragment : ABFragment(R.layout.fragment_map), UserLocationListener, MenuProvider {

    companion object {
        private val LOG_TAG = MapFragment::class.java.simpleName

        private const val TRACKING_SCREEN_NAME = "Map"

        @JvmStatic
        fun newInstance(poim: POIManager) = newInstance(
            LocationUtils.getNewLocation(poim.lat, poim.lng),
            poim.poi.uuid,
            poim.poi.dataSourceTypeId
        )

        @JvmStatic
        @JvmOverloads
        fun newInstance(
            optInitialLocation: Location? = null,
            optSelectedUUID: String? = null,
            optIncludeTypeId: Int? = null,
        ): MapFragment {
            return MapFragment().apply {
                arguments = newInstanceArgs(optInitialLocation, optSelectedUUID, optIncludeTypeId)
            }
        }

        @JvmStatic
        fun newInstanceArgs(poim: POIManager) = newInstanceArgs(
            LocationUtils.getNewLocation(poim.lat, poim.lng),
            poim.poi.uuid,
            poim.poi.dataSourceTypeId
        )

        @JvmStatic
        @JvmOverloads
        fun newInstanceArgs(
            optInitialLocation: Location? = null,
            optSelectedUUID: String? = null,
            optIncludeTypeId: Int? = null,
        ) = bundleOf(
            MapViewModel.EXTRA_INITIAL_LOCATION to optInitialLocation,
            MapViewModel.EXTRA_SELECTED_UUID to optSelectedUUID,
            MapViewModel.EXTRA_INCLUDE_TYPE_ID to (optIncludeTypeId ?: MapViewModel.EXTRA_INCLUDE_TYPE_ID_DEFAULT),
        )
    }

    override fun getLogTag(): String = LOG_TAG

    override fun getScreenName(): String = TRACKING_SCREEN_NAME

    private val viewModel by viewModels<MapViewModel>()
    private val attachedViewModel
        get() = if (isAttached()) viewModel else null

    private var binding: FragmentMapBinding? = null

    @Inject
    lateinit var dataSourcesRepository: DataSourcesRepository

    @Inject
    lateinit var locationPermissionProvider: LocationPermissionProvider

    private val mapListener = object : MapViewController.MapListener {

        override fun onMapClick(position: LatLng) {
            // DO NOTHING
        }

        override fun onCameraChange(latLngBounds: LatLngBounds) {
            attachedViewModel?.onCameraChange(latLngBounds) {
                mapViewController.getBigCameraPosition(activity, 1.0f)
            }
        }

        override fun onMapReady() {
            attachedViewModel?.poiMarkers?.value?.let {
                mapViewController.clearMarkers()
                mapViewController.addMarkers(it)
                mapViewController.showMap(view)
            }
        }
    }

    private val mapViewController: MapViewController by lazy {
        MapViewController(
            logTag,
            null, // DO NOTHING (not linked with list adapter)
            mapListener,
            true,
            true,
            true,
            false,
            false,
            false,
            64,
            0,
            false,
            true,
            true,
            false,
            true,
            this.dataSourcesRepository
        ).apply {
            logTag = this@MapFragment.logTag
            setLocationPermissionGranted(locationPermissionProvider.allRequiredPermissionsGranted(requireContext()))
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mapViewController.apply {
            setDataSourcesRepository(dataSourcesRepository)
            onAttach(requireActivity())
            setLocationPermissionGranted(locationPermissionProvider.allRequiredPermissionsGranted(context))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mapViewController.onCreate(savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        (requireActivity() as MenuHost).addMenuProvider(
            this, viewLifecycleOwner, Lifecycle.State.RESUMED
        )
        this.mapViewController.onViewCreated(view, savedInstanceState)
        binding = FragmentMapBinding.bind(view).apply {
            // DO NOTHING
        }
        viewModel.initialLocation.observe(viewLifecycleOwner) { location ->
            location?.let {
                mapViewController.setInitialLocation(it)
                viewModel.onInitialLocationSet()
            }
        }
        viewModel.selectedUUID.observe(viewLifecycleOwner) {
            it?.let {
                mapViewController.setInitialSelectedUUID(it)
                viewModel.onSelectedUUIDSet()
            }
        }
        viewModel.deviceLocation.observe(viewLifecycleOwner) {
            context?.let { context ->
                mapViewController.setLocationPermissionGranted(locationPermissionProvider.allRequiredPermissionsGranted(context))
            }
            mapViewController.onDeviceLocationChanged(it)
        }
        viewModel.filterTypeIds.observe(viewLifecycleOwner) {
            abController?.setABTitle(this, getABTitle(context), true)
        }
        viewModel.typeMapAgencies.observe(viewLifecycleOwner) {
            viewModel.resetLoadedPOIMarkers()
        }
        viewModel.poiMarkersTrigger.observe(viewLifecycleOwner) {
            // DO NOTHING
        }
        viewModel.loaded.observe(viewLifecycleOwner) {
            if (it == false) {
                mapViewController.showLoading()
            } else if (it == true) {
                mapViewController.hideLoading()
            }
        }
        viewModel.poiMarkers.observe(viewLifecycleOwner) {
            mapViewController.clearMarkers() // new types -> RESET
            it?.let { poiMarkers ->
                mapViewController.addMarkers(poiMarkers)
                mapViewController.showMap(view)
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        mapViewController.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onDetach() {
        super.onDetach()
        this.mapViewController.onDetach()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        mapViewController.onConfigurationChanged(newConfig)
    }

    override fun onResume() {
        super.onResume()
        mapViewController.onResume()
        mapViewController.showMap(view)
        (activity as? MTActivityWithLocation)?.let { onUserLocationChanged(it.lastLocation) }
    }

    override fun onUserLocationChanged(newLocation: Location?) {
        attachedViewModel?.onDeviceLocationChanged(newLocation)
    }

    override fun onPause() {
        super.onPause()
        mapViewController.onPause()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapViewController.onLowMemory()
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.menu_map, menu)
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        return if (menuItem.itemId == R.id.menu_filter) {
            showMenuFilterDialog()
        } else false // not handled
    }

    private fun showMenuFilterDialog(): Boolean {
        val filterTypeIds = attachedViewModel?.filterTypeIds?.value ?: return false
        val activity = requireActivity()
        val typeNames = mutableListOf<CharSequence>()
        val checked = mutableListOf<Boolean>()
        val typeIds = mutableListOf<Int>()
        val selectedItems = mutableSetOf<Int>()
        attachedViewModel?.mapTypes?.value?.forEach { type ->
            typeIds.add(type.id)
            typeNames.add(getString(type.poiShortNameResId))
            checked.add(filterTypeIds.isEmpty() || filterTypeIds.contains(type.id))
        }
        val checkedItems = BooleanArray(checked.size)
        for (c in checked.indices) {
            checkedItems[c] = checked[c]
            if (checkedItems[c]) {
                selectedItems.add(c)
            }
        }
        MTDialog.Builder(activity)
            .setTitle(R.string.menu_action_filter)
            .setMultiChoiceItems(typeNames.toTypedArray(), checkedItems) { _, which, isChecked ->
                selectedItems.apply { if (isChecked) add(which) else remove(which) }
            }
            .setPositiveButton(android.R.string.ok) { dialog, _ ->
                applyNewFilter(typeIds, selectedItems)
                dialog.dismiss()
            }
            .setNegativeButton(android.R.string.cancel) { dialog, _ ->
                dialog.dismiss()
            }
            .setCancelable(true)
            .create()
            .show()
        return true // handled
    }

    private fun applyNewFilter(availableTypeIds: List<Int>, selectedTypesIdx: Set<Int>) {
        viewModel.saveFilterTypeIdsPref(
            if (selectedTypesIdx.isEmpty() || selectedTypesIdx.size == availableTypeIds.size) { // ALL SELECTED
                null
            } else { // NOT ALL SELECTED
                availableTypeIds.filterIndexed { idx, _ ->
                    selectedTypesIdx.contains(idx)
                }
            }
        )
    }

    override fun getABBgColor(context: Context?) = Color.TRANSPARENT

    override fun getABTitle(context: Context?): CharSequence? {
        return context?.let { makeABTitle(it) } ?: super.getABTitle(null)
    }

    private fun makeABTitle(context: Context): CharSequence {
        return (attachedViewModel?.filterTypeIds?.value?.let { it.ifEmpty { null } } // empty = all
            ?.mapNotNull { typeId ->
                DataSourceType.parseId(typeId)?.allStringResId?.let { context.getString(it) }
            } ?: listOf(context.getString(R.string.all)))
            .joinToString(prefix = "${context.getString(R.string.map)} (", separator = ", ", postfix = ")")
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mapViewController.onDestroyView()
        binding = null
    }

    override fun onDestroy() {
        super.onDestroy()
        mapViewController.onDestroy()
    }
}